### ISSUE TITLE
Update to Fody Version 2

### ISSUE THREAD

**Issue by SimonCropp** [csnemes/tracer#46]
Fody Version 2 is out and you will need to do an update for your addin to be usable with it

Specifically:

 * Update to the newest Fody.Cecil nuget
 * Change the min dependency range for Fody to be at least 2.0

 Please ping me if u need help

**Comment by Diagonactic** [comment]
Yep - same issue here - I've got a solution that depends on this and a weaver I wrote that I referenced version 2 in.  I **think** it'll be an easy enough thing to resolve, so I've cloned and am updating references.  If I get it sorted out and it works, I'll send a PR to the project.

**Comment by csnemes** [comment]
Hi, I'm on it, it just seems to be a bit more tricky than I thought as it won't compile due to some missing properties. Anyhow I'll try to fix it today.

**Comment by csnemes** [comment]
My add-in works fine after porting and adopting to the new fody and cecil libraries except that the debug information is messed-up (at least in visual studio). You cannot set breakpoints on some lines of the modified methods and some breakpoints move some lines away (even to different methods) when you start the weaved program in debug mode. Seems like the sequence points are out of sync, but I don't temper them, even a quick look in the debugger shows that their are pointing to the right instructions. Have you experienced anything like this?

**Comment by Diagonactic** [comment]
Hi @csnemes - Excellent!!

I poked at it last night and a bit this afternoon.  The issue is that the new Mono.Cecil, which is required by Fody 2.0 eliminated the SequencePoint property, which it appears in CecilExtensions.cs is being retrieved and later set for an instruction.

I changed `InsertAtTheBeginning` in that class to get the sequence point using the guidance provided by the Mono.Cecil project:

```c#
var enteringSeqPoint = body.Method.DebugInformation.GetSequencePoint(body.Instructions[0]);
```

But I can't find a way to set the sequence point as is done a couple of lines below that with:

```c#
body.Instructions[0].SequencePoint = enteringSeqPoint;
```

They've also changed the VariableDefinition class to no longer contain a name.  According to the documentation, this is because:

> The property VariableReference.Name has been removed. This has never been represented in the metadata, and is actually incorrect. From a debug information perspective, one variable slot in IL could have different semantic names in the source.

I ran into a similar problem with that bit of code.  They've provided a way to get at the name (if possible), however, no way to create a new declaration with a name specified that I could find.  To get the name, I created a new method:

```c#
private static bool NameMatchesVariableDefinition(MethodDefinition method, VariableDefinition definition, string name)
        {
            string variableName;
            return method.DebugInformation.TryGetName(definition, out variableName) && variableName.Equals(name);
        }
```

And used it in `GetVariable` and `GetOrDeclareVariable`, replacing the first line of those methods with:

```c#
var variable = body.Variables.FirstOrDefault(p => NameMatchesVariableDefinition(body.Method, p, name));
```

But in `GetOrDeclareVariable`, the declare code:

```c#
variable = new VariableDefinition(name, type);
```

... does not have an equivalent way to *set* the name and the constructor now only accepts the type parameter.

Those two spots are the only ones that I am unsure of how to resolve.  I'd imagine that's why you're seeing issues with breakpoints not being able to be set, since the PDB is probably getting out of sync with the code since it doesn't appear to be possible to set the `SequencePoint` in the `InsertAtTheBeginning` method.  I'm not sure what impact not setting the variable name has on the project (the GetOrDeclare method had 15 references, so it's used plenty), but considering the new Mono.Cecil documentation, finding a workaround that doesn't involve (or minimizes) the need for variable names is probably a good idea if they're not set properly (unless this is being used to get names that *are* reliable to retrieve, in which case, being able to set the names is going to still be a problem).

Not sure if what I've provided helps you out at all, but if you have more information, I'm happy to try tackle this with you.  It's a *great* project ... one of those things that once you have it it's hard to live without ... it's amazing how much cleaner code becomes when you can eliminate almost all of the logging pollution (or how easily instrumented code becomes when you can automagically log everything with a single FodyWeaver directive).

**Comment by csnemes** [comment]
Hi,
Thank you for your great work. I updated the structure of the code a little bit to better fit in the new way of doing things, so I've got rid of the GetOrDeclare and saved the variable defs in local fields created with Declare. In general it works like a charm except for that nasty debug problem. I updated the dev branch with the new code but I don't want to release it until the debugging problem is fixed. Regarding this debug issue it seems that either I'm doing something in a wrong way or there's some issue with cecil. I've managed to create a small test app to reproduce the problem so it has nothing to do with fody as I can see the problem when I inject some extra calls into a method using only cecil. The strange thing is that dotPeek shows the correct matching between IL and source yet Visual Studio 2015 (and 2017) gets crazy and you're not able to set some valid breakpoints (they appear as empty circle and VS complains about them). I'll upload the test project to github and I'd like to ask you to check if you can reproduce the problem or spot any errors in the way I do the injection. Thanks a lot, once again.

**Comment by csnemes** [comment]
The repro can be found at https://github.com/csnemes/CecilTest

**Comment by Diagonactic** [comment]
Awesome! I probably won't get to take a look at it this afternoon due to some appointments I have, but I should have some time tomorrow.

The biggest concern I had was around the variable definitions and it looks like you've got that solved (it looked like the harder of the two to resolve and I wasn't sure what purpose that served, so I wasn't sure what I could change without breaking things).  With that out of the way, it's just a matter of setting the sequence point ... which it looks like Mono.Cecil doesn't allow any longer ... but even without it, it should be doable.  At a minimum, the PDB format is now documented, so it could be done "the hard way", but I have a hunch there's a work-around that won't require rolling a PDB parser (or, maybe I'm just **really** hoping that's the case!).

Thanks, sir - will hop in there and see if I can throw any help your way.

**Comment by csnemes** [comment]
I asked JB (author of Cecil) about this issue. You can follow the thread at https://groups.google.com/forum/#!topic/mono-cecil/tNOeRiTBtXo

**Comment by Diagonactic** [comment]
I ran the test program you provided through cvdump.exe (Microsoft's PDB dump tool) and ran into some differences that *I'm guessing* are related to Mono.Cecil.  It's entirely possible that some of these pieces are being removed due to the `Optimize` call, which may simply be doing exactly what it's supposed to do.  It's a bit of tall weeds for me, so I'm not sure but I'll be studying this a bit this weekend to see if I can figure it out.

For now, I thought I'd provide the notes from my analysis so that (a) I actually take notes and (b) on the off chance that you notice something I am not seeing.

Here are the sections of cvdump.exe that are different between the two files.

### Symbols - Before Rewriting
```
*** SYMBOLS

** Module: "CecilTest.Program" from "466D5406"


(000004) S_GMANPROC: [0001:00000000], Cb: 00000024, Token: 06000001, Main
         Parent: 00000000, End: 000000F4, Next: 00000000
         Debug start: 00000000, Debug end: 00000000
         Return Reg: ???(0x0000)

(000034)  S_BLOCK32: [0001:00000000], Cb: 00000024, 
          Parent: 00000004, End: 000000BC

(00004C)   S_UNAMESPACE: USystem
(000058)   S_UNAMESPACE: USystem.Collections.Generic
(000078)   S_UNAMESPACE: USystem.Linq
(00008C)   S_UNAMESPACE: USystem.Text
(0000A0)   S_UNAMESPACE: USystem.Threading.Tasks

(0000BC)  S_END

(0000C0)  S_OEM: {C6EA3FC9-59B3-49D6-BC25-0902BBABB460}, Type:     T_NOTYPE(0000)

(0000F4) S_END
```

### Symbols - After Rewriting
Note that in this case there are far less.  In addition, the End value is much lower.
```
*** SYMBOLS

** Module: "CecilTest.Program" from "466D5406"


(000004) S_GMANPROC: [0001:00000000], Cb: 00000034, Token: 06000001, Main
         Parent: 00000000, End: 00000034, Next: 00000000
         Debug start: 00000000, Debug end: 00000000
         Return Reg: ???(0x0000)

(000034) S_END
```

### Segment Map - Before
```
*** SEGMENT MAP

Sec  flags  ovl   grp   frm sname cname    offset    cbSeg
 01  010d  0000  0000  0001  ffff  ffff  00000000 00000024
 02  0208  0000  0000  0000  ffff  ffff  00000000 ffffffff
```

### Segment Map - After
If my understanding of this is correct, this looks OK.  We've added some instructions and based on the Debug output from the test, it seems to line up with the IL changes.
```
*** SEGMENT MAP

Sec  flags  ovl   grp   frm sname cname    offset    cbSeg
 01  010d  0000  0000  0001  ffff  ffff  00000000 00000034
 02  0208  0000  0000  0000  ffff  ffff  00000000 ffffffff
```

### Section Headers - Before
```
*** SECTION HEADERS


SECTION HEADER #1
         name
      24 virtual size
       0 virtual address
      24 size of raw data
       0 file pointer to raw data
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
       0 flags
         (no align specified)
```

### Section Headers - After
Same with this - seems to match up with what I'd expect from the debug output.
```
*** SECTION HEADERS


SECTION HEADER #1
         name
      34 virtual size
       0 virtual address
      34 size of raw data
       0 file pointer to raw data
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
       0 flags
         (no align specified)
```

### Thoughts
Of the information from the .PDB, the only thing that doesn't quite add up for me is the symbols section.  It seems like it's just *wiped out*.  If it were optimization, I'd kind of expect this to simply have less data, but not for all of it to be just ... *gone*.  So I'm thinking something is getting b0rked there.  Now, whether or not this has *anything at all* to do with what's going wrong is not something I can speak intelligently about, yet.  Hopefully I'll understand this a little more, later, and can do some more experimenting.

**Comment by csnemes** [comment]
Thanks for the great analysis. There's certainly something fishy with the modified PDB file. I'm going to spend some time on this issue at the weekend, see if I can get closer to the solution. Thanks again for your work.

**Comment by jesuissur** [comment]
@csnemes JB said he has fixed in Cecil's master
https://groups.google.com/d/msg/mono-cecil/tNOeRiTBtXo/88yVZyDDAQAJ

**Comment by jesuissur** [comment]
I've run the cvdump with the latest cecil build and the symbols section now looks ok

![image](https://cloud.githubusercontent.com/assets/136296/26168273/819f816e-3b07-11e7-9f50-071cdda5deb5.png)

Dump with the latest cecil

```
*** SYMBOLS

** Module: "CecilTest.Program" from "466D5406"


(000004) S_GMANPROC: [0001:00000000], Cb: 00000034, Token: 06000001, Main
         Parent: 00000000, End: 000000F0, Next: 00000000
         Debug start: 00000000, Debug end: 00000000
         Return Reg: ???(0x0000)

(000034)  S_BLOCK32: [0001:00000000], Cb: 00000034,
          Parent: 00000004, End: 000000BC

(00004C)   S_UNAMESPACE: USystem
(000058)   S_UNAMESPACE: USystem.Collections.Generic
(000078)   S_UNAMESPACE: USystem.Linq
(00008C)   S_UNAMESPACE: USystem.Text
(0000A0)   S_UNAMESPACE: USystem.Threading.Tasks

(0000BC)  S_END

(0000C0)  S_OEM: {C6EA3FC9-59B3-49D6-BC25-0902BBABB460}, Type:     T_NOTYPE(0000)

(0000F0) S_END


*** SEGMENT MAP

Sec  flags  ovl   grp   frm sname cname    offset    cbSeg
 01  010d  0000  0000  0001  ffff  ffff  00000000 00000034
 02  0208  0000  0000  0000  ffff  ffff  00000000 ffffffff

*** SECTION HEADERS


SECTION HEADER #1
         name
      34 virtual size
       0 virtual address
      34 size of raw data
       0 file pointer to raw data
       0 file pointer to relocation table
       0 file pointer to line numbers
       0 number of relocations
       0 number of line numbers
       0 flags
         (no align specified)
```

**Comment by csnemes** [comment]
Great. Thanks for your help. I've updated the dependency to Fody 2.0.9 and released a new version (Tracer 1.6.)
